src/merkle/merkle_tree.rs

```diff
@@ -158,6 +160,47 @@ impl MerkleTree {
 
         Ok(())
     }
+
+    /// An iterator over every inner node in the tree. The iterator order is unspecified.
+    pub fn inner_nodes(&self) -> MerkleTreeNodes<'_> {
+        MerkleTreeNodes {
+            nodes: &self.nodes,
+            index: 1, // index 0 is just padding, start at 1
+        }
+    }
+}
+
+// ITERATORS
+// ================================================================================================
+
+/// An iterator over every inner node of the [MerkleTree].
+///
+/// Use this to extract the data of the tree, there is no guarantee on the order of the elements.
+pub struct MerkleTreeNodes<'a> {
+    nodes: &'a Vec<Word>,
+    index: usize,
+}
+
+impl<'a> Iterator for MerkleTreeNodes<'a> {
+    type Item = InnerNodeInfo;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.index * 2 < self.nodes.len() {
```

Comment:
This shouldn't really be a problem, but I wonder if changing this to `self.index2 < self.nodes.len() / 2` might be better. The reasons is that it is theoretically possible (though not realistic) for this structure to have close to 2^64 nodes, and thus, 2 * index may overflow.

Reply:
good point! changing it